vcd/resource_vcd_vapp_vm.go

```diff
@@ -1596,6 +1640,55 @@ func readNetworks(vm govcd.VM, vapp govcd.VApp) ([]map[string]interface{}, error
 
 		nets = append(nets, singleNIC)
 	}
+
+	vmStatus, err := vm.GetStatus()
+	if err != nil {
+		return nil, fmt.Errorf("unablet to check if VM is powered on: %s", err)
+	}
+
+	// If at least one`network_dhcp_wait_seconds` was defined
+	if maxDhcpWaitSeconds, ok := d.GetOk("network_dhcp_wait_seconds"); ok && vmStatus == "POWERED_ON" {
+		maxDhcpWaitSecondsInt := maxDhcpWaitSeconds.(int)
+
+		// lookup NIC indexes which have DHCP enabled
+		dhcpNicIndexes := getVmNicIndexesWithDhcpEnabled(vm.VM.NetworkConnectionSection)
+		log.Printf("[DEBUG] [VM read] '%s' DHCP is used on NICs %v with wait time '%d seconds'",
+			vm.VM.Name, dhcpNicIndexes, maxDhcpWaitSecondsInt)
+		if len(dhcpNicIndexes) == 0 {
+			_, _ = fmt.Fprint(getTerraformStdout(), "INFO: Using 'network_dhcp_wait_seconds' only "+
+				"makes sense if at least one NIC is using 'ip_allocation_mode=DHCP'")
+		}
+
+		if len(dhcpNicIndexes) > 0 { // at least one NIC uses DHCP for IP allocation mode
+			log.Printf("[DEBUG] [VM read] '%s' waiting for DHCP IPs up to '%d' seconds on NICs %v",
+				vm.VM.Name, maxDhcpWaitSeconds, dhcpNicIndexes)
+
+			start := time.Now()
+			nicIps, timeout, err := vm.WaitForDhcpIpByNicIndexes(dhcpNicIndexes, maxDhcpWaitSecondsInt, true)
+			if err != nil {
+				return nil, fmt.Errorf("unable to to lookup DHCP IPs for VM NICs '%v': %s", dhcpNicIndexes, err)
+			}
+
+			if timeout {
+				log.Printf("[DEBUG] [VM read] VM %s timed out waiting %d seconds "+
+					"to report DHCP IPs. You may want to increase 'network_dhcp_wait_seconds' or ensure "+
+					"your DHCP settings are correct.\n", vm.VM.Name, maxDhcpWaitSeconds)
+				_, _ = fmt.Fprintf(getTerraformStdout(), "WARNING: VM %s timed out waiting %d seconds "+
```

Comment:
I tried to hit the timeout. It's nice that we get the WARNING inside the console. However, the further plan ended up running with a bunch of errors, because NAT+FW rules tried to use empty IP. Do we have any way to protect against that or workaround in the `.tf` config itself?

```
vcd_vapp_vm.demo_vm_wordpress: Still creating... [4m50s elapsed]
WARNING: VM demo-vm-wordpress timed out waiting 60 seconds to report DHCP IPs. You may want to increase 'network_dhcp_wait_seconds' or ensure your DHCP settings are correct.
vcd_vapp_vm.demo_vm_wordpress: Creation complete after 4m52s [id=urn:vcloud:vm:5322ba15-9fc6-4e56-a193-6072f6c66129]
vcd_nsxv_dnat.demo_dnat_wordpress_ssh: Modifying... [id=196613]
vcd_nsxv_snat.web: Modifying... [id=196612]
vcd_nsxv_dnat.demo_dnat_wordpress_https: Modifying... [id=196614]
vcd_nsxv_firewall_rule.demo-nsxv-fw-rule-outbound: Modifying... [id=131079]
vcd_nsxv_dnat.demo_dnat_wordpress_https: Still modifying... [id=196614, 10s elapsed]
vcd_nsxv_firewall_rule.demo-nsxv-fw-rule-outbound: Still modifying... [id=131079, 10s elapsed]
vcd_nsxv_firewall_rule.demo-nsxv-fw-rule-outbound: Still modifying... [id=131079, 20s elapsed]
vcd_nsxv_firewall_rule.demo-nsxv-fw-rule-outbound: Modifications complete after 25s [id=131079]

Error: unable to update NAT rule with ID 196614: error while updating NAT rule : vShield Edge Invalid IP Address input /32 for field translatedAddress. (API error: 15012)

  on xendi.tf line 393, in resource "vcd_nsxv_dnat" "demo_dnat_wordpress_https":
 393: resource "vcd_nsxv_dnat" "demo_dnat_wordpress_https" {



Error: unable to update NAT rule with ID 196613: error while updating NAT rule : vShield Edge Invalid IP Address input /32 for field translatedAddress. (API error: 15012)

  on xendi.tf line 411, in resource "vcd_nsxv_dnat" "demo_dnat_wordpress_ssh":
 411: resource "vcd_nsxv_dnat" "demo_dnat_wordpress_ssh" {



Error: unable to update NAT rule with ID 196612: error while updating NAT rule : vShield Edge Invalid IP Address input /32 for field originalAddress. (API error: 15012)

  on xendi.tf line 430, in resource "vcd_nsxv_snat" "web":
 430: resource "vcd_nsxv_snat" "web" {
```

Reply:
The first thing that comes to my mind is actually returning an **error** during READ instead of **warning**. But there are a few problems and I believe this isn't good enough:
* READ is part of CREATE/UPDATE so creation could fail because of too small value.
* Because READ accesses the configuration from statefile, even changing the value to a higher one (say from 1 second to 300 seconds) would still at first try to READ with 1 second (because plan/apply triggers READ before UPDATE). If it fails immediately then the user never gets to 300 seconds.

Reply:
Yes, looks like current behavior is actually correct. Given that VM _does_ get created even though without IP address, we _shouldn't_ report an error. In a sense, we're missing a special state in the Terraform state machine to model such incomplete executions. All in all, let's leave current implementation intact.